src/frontend/json.rs

```diff
@@ -0,0 +1,833 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! Module defining the logic for compiling the deserialized json file into
+//! the IR. (Intermediate Representation)
+//!
+//! It also defines some of the objects that a JSON seccomp filter is deserialized into:
+//! [`JsonFilter`](struct.JsonFilter.html),
+//! [`JsonRule`](struct.JsonRule.html),
+//! [`JsonCondition`](struct.JsonCondition.html).
+//
+//! The rest of objects are deserialized directly into the IR :
+//! [`SeccompCondition`](struct.SeccompCondition.html),
+//! [`SeccompAction`](enum.SeccompAction.html),
+//! [`SeccompCmpOp`](enum.SeccompCmpOp.html),
+//! [`SeccompCmpArgLen`](enum.SeccompCmpArgLen.html).
+
+use std::collections::{BTreeMap, HashMap};
+use std::convert::{TryFrom, TryInto};
+use std::fmt;
+use std::io::Read;
+use std::result;
+
+use crate::backend::{
+    Error as BackendError, SeccompAction, SeccompCmpArgLen, SeccompCmpOp, SeccompCondition,
+    SeccompFilter, SeccompRule, TargetArch,
+};
+use crate::syscall_table::SyscallTable;
+use serde::de::{self, Error as _, MapAccess, Visitor};
+use serde::{Deserialize, Deserializer};
+
+type Result<T> = result::Result<T, Error>;
+
+/// Error compiling JSON into IR.
+#[derive(Debug)]
+pub enum Error {
+    /// Backend error creating the `SeccompFilter` IR.
+    Backend(BackendError),
+    /// Error deserializing JSON.
+    SerdeJson(serde_json::Error),
+    /// Invalid syscall name for the given arch.
+    SyscallName(String, TargetArch),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match self {
+            Backend(error) => write!(f, "{}", error),
+            SerdeJson(error) => {
+                write!(f, "Error parsing Json: {}", error)
+            }
+            SyscallName(syscall_name, arch) => write!(
+                f,
+                "Invalid syscall name: {} for given arch: {:?}.",
+                syscall_name, arch
+            ),
+        }
+    }
+}
+
+/// Deserializable object that represents the top-level map of Json Filters.
+// Need the 'newtype' pattern so that we can implement a custom deserializer.
+pub(crate) struct JsonFilterMap(pub HashMap<String, JsonFilter>);
+
+// Implement a custom deserializer, that returns an error for duplicate thread keys.
```

Comment:
You can do this with `serde_with`: https://docs.rs/serde_with/1.10.0/serde_with/rust/maps_duplicate_key_is_error/index.html.

Reply:
I was also thinking about that, but `serde_with` pulls even more dependencies that we can otherwise avoid.

Reply:
yep, I considered that as well, but I think it's best to avoid dependencies that we can otherwise work around quite easily